### DISCUSSION TITLE
Possible to exclude keywords or attributes being added to search index?

### DISCUSSION THREAD

**Discussion by jamesmacwhite** [craftcms/cms#11575] [Ideas] [answered]
Hi,

I've migrated a large Craft CMS 2 site and because nearly all of the custom fields have been created in Craft CMS 2, they are all default on for being searchable, this doesn't make sense for the majority of the fields i.e. numeric, lightswitch etc, so I'm currently in batches turning off the searchable setting and pushing these up via project config to reduce the search index table down in size. I am having to do it batches due to the field changes taking a while to apply sometimes, I assume because search index data is being removed.

It did however make me think if there was an event to hook into that would prevent adding certain data into the search index keywords/attributes regardless.

E.g. if a keyword is less than or equal to 3 characters or being a numeric value.

Another developer @ostark has created a plugin that does this with the search index: https://github.com/ostark/craft-relax/blob/main/src/SearchIndex/InsertFilter/UselessKeywords.php

It would be potentially useful to have an event that a module could hook into that is able to return (or skip) if a value matches something like being less than a certain amount of characters or if it's a numeric value i.e. 0 or 1 as these will probably not really be useful for searching ever. Of course this is possibly opinionated, so that's why having the event itself to hook into would be useful to create these custom rules.

Thanks,

James

**Comment by brandonkelly** [comment, marked as answer]
I’ve just added a new `EVENT_BEFORE_INDEX_KEYWORDS` event to the Search service, for Craft 4.2. It can be used to centrally modify search keywords before they’re indexed.

```php
use craft\events\IndexKeywordsEvent;
use craft\services\Search;
use yii\base\Event;

Event::on(
    Search::class,
    Search::EVENT_BEFORE_INDEX_KEYWORDS,
    function(IndexKeywordsEvent $event) {
        $event->keywords = 'foo bar baz';
    }
);
```

Note that it’s important that the `keywords` remain a string of words separated by single spaces. Otherwise LIKE/exact match searching will break.

**Reply by jamesmacwhite** [reply]
@brandonkelly Amazing, thank you!

**Reply by brandonkelly** [reply]
Craft 4.2.0 is out now with that event ✨

**Reply by d--j** [reply]
Could we also add the possibility to conditionally skip the creation of the whole search index DB row?

Something like

https://github.com/craftcms/cms/blob/3db7fb231049166f102aa75a5a23b52a8762b7d1/src/services/Search.php#L340-L349


```php
        if ($this->hasEventHandlers(self::EVENT_BEFORE_INDEX_KEYWORDS)) {
            $event = new IndexKeywordsEvent([
                'element' => $element,
                'attribute' => $attribute,
                'fieldId' => $fieldId,
                'keywords' => $keywords,
            ]);
            $this->trigger(self::EVENT_BEFORE_INDEX_KEYWORDS, $event);
            if ($event->handled) {
                return;
            }
            $keywords = $event->keywords;
        }
```

**Reply by brandonkelly** [reply]
@d--j Can you create a separate discussion for that?
